File: data_structures/binarySearch_tree/Java/BinarySearchTree.java
public class BinarySearchTree {
	private int value;
	private BinarySearchTree left;
	private BinarySearchTree right;

	public BinarySearchTree(int value) {
		this.value = value;
		this.left = null;
		this.right = null;
	}

	public String printInorder(BinarySearchTree root) {
		String result = "";
		
		if(root.left != null) {
			result = result + root.printInorder(root.left);
		}

		result = result + root.value;

		if(root.right != null) {
			result = result + root.printInorder(root.right);
		}

		return result;
	}

	public int height(BinarySearchTree root) {
		if(root == null) {
			return 0;
		}

		int left = 1 + root.height(root.left);
		int right = 1 + root.height(root.right);

		if(left > right) {
			return left;
		} else {
			return right;
		}
	}

	public String printPreorder(BinarySearchTree root) {
		String result = "";
		
		result = result + root.value;
		
		if(root.left != null) {
			result = result + root.printPreorder(root.left);
		}

		if(root.right != null) {
			result = result + root.printPreorder(root.right);
		}

		return result;
	}

	public String printPostorder(BinarySearchTree root) {
		String result = "";
		
		if(root.left != null) {
			result = result + root.printPostorder(root.left);
		}

		if(root.right != null) {
			result = result + root.printPostorder(root.right);
		}

		result = result + root.value;
		
		return result;
	}

	public BinarySearchTree makeNode(int value) {
		BinarySearchTree node = new BinarySearchTree(value);
		return node;
	}

	public int minimum(BinarySearchTree root) {
		if(root != null) {
			while(root.left != null) {
				root = root.left;
			}
			return root.value;
		}
		return -1;
	}

	public BinarySearchTree delete(BinarySearchTree root, int value) {
		if(root == null) {
			return root;
		}

		if(root.value > value) {
			root.left = root.delete(root.left, value);
		} else if(root.value < value) {
			root.right = root.delete(root.right, value);
		} else {
			if(root.left != null && root.right != null) {
				int tmp = root.minimum(root.right);
				root.value = tmp;
				root.right = root.delete(root.right, tmp);
			} else {
				BinarySearchTree tmp = root;
				if(root.left != null) {
					root = root.left;
				} else {
					root = root.right;
				}
			}
		}

		return root;
	}

	public BinarySearchTree addNode(BinarySearchTree root, int value) {
		if(root == null) {
			root = root.makeNode(value);
		} else if(root.value == value) {
			return root;
		} else if(root.value > value) {
			if(root.left == null) {
				root.left = root.makeNode(value);
			} else {
				root.left = root.addNode(root.left, value);
			}
		} else if(root.value < value) {
			if(root.right == null) {
				root.right = root.makeNode(value);
			} else {
				root.right = root.addNode(root.right, value);
			}
		}

		return root;
	}

	public static void main(String []args) {
		BinarySearchTree tree = null;
		tree = new BinarySearchTree(5);
		tree = tree.addNode(tree, 3);
		tree = tree.addNode(tree, 2);
		tree = tree.addNode(tree, 4);
		tree = tree.addNode(tree, 7);
		tree = tree.addNode(tree, 6);
		tree = tree.addNode(tree, 8);

		System.out.println("Inorder: " + tree.printInorder(tree));
		System.out.println("Postorder: " + tree.printPostorder(tree));
		System.out.println("Preorder: " + tree.printPreorder(tree));
		System.out.println("Height: " + tree.height(tree));

		tree = tree.delete(tree, 5);

		System.out.println("Inorder: " + tree.printInorder(tree));
	}
}

File: data_structures/binarySearch_tree/clojure/BinarySearchTree.clj
(ns Algorithms.BinarySearchTree)

(defrecord BinarySearchTree [val left right path])

(defn populate-parents
  "Return t with the parent field filled in"
  ([tree] (populate-parents tree []))
  ([tree path]
     (cond (leaf? tree) (BinarySearchTree. (:val tree) nil nil path)
           (= (:left tree) nil) (BinarySearchTree. (:val tree)
                                                   nil
                                                   (populate-parents (:right tree) (conj path :right))
                                                   path)
           (= (:right tree) nil) (BinarySearchTree. (:val tree)
                                                    (populate-parents (:left tree) (conj path :left))
                                                    nil
                                                    path)
           :else (BinarySearchTree. (:val tree)
                                    (populate-parents (:left tree) (conj path :left))
                                    (populate-parents (:right tree) (conj path :right))
                                    path))))

(defn walk-tree-inorder [t node-fn]
  "Visit each node of the tree in order, calling node-fn on each"
  (when t
    (do (walk-tree-inorder (:left t) node-fn)
        (node-fn t)
        (walk-tree-inorder (:right t) node-fn))))

(defn walk-tree-preorder [t node-fn]
    "Visit each node of the tree in preorder, calling node-fn on each"
  (when t
    (do (node-fn t)
        (walk-tree-preorder (:left t) node-fn)
        (walk-tree-preorder (:right t) node-fn))))

(defn walk-tree-postorder [t node-fn]
      "Visit each node of the tree in postorder, calling node-fn on each"
  (when t
    (do (walk-tree-postorder (:left t) node-fn)
        (walk-tree-postorder (:right t) node-fn)
        (node-fn t))))

(defn print-tree [t walk-fn]
  "Print each value in the tree, with the order determined by walk-fn"
  (walk-fn t #(println (:val %))))

(defn search [t val]
  "Return the node that contains val"
  (if (or (= t nil) (= (:val t) val))
    (loop [tree t]
      (let [branch (if (< (:val tree) val)
                     (:left tree)
                     (:right tree))]
        (recur branch)))))

(defn minimum [t]
  "Return the node contailing the smallest value"
  (if (= (:left t) nil)
    t
    (recur (:left t))))

(defn maximum [t]
  "Return the node contailing the largest value"
  (if (= (:right t) nil)
    t
    (recur (:right t))))

(defn leaf? [t]
  "True if t is a leaf node, false otherwise"
  (and (= (:left t) nil)
       (= (:right t) nil)))

(defn successor [t]
  (minimum (:right t)))

(defn follow-directions
  "Return the node from tree arrived at by following the directions in path"
  ([tree path] (if (not (seq path))
                 nil
                 (follow-directions tree (rest path) ((first path) tree))))
  ([tree path acc] (if (not (seq path))
                     acc
                     (follow-directions tree (rest path) ((first path) tree)))))